syntax: better error message for missing repetition quantifier

christianrondeau opened the pull request:
Hi, first contribution to the Rust ecosystem, hopefully I'll get it right.

This is to solve @CAD97 issue https://github.com/rust-lang/regex/issues/545 which raised an uncleared error message when repetition quantifiers were not followed by a valid decimal (other languages would treat the `{}` as characters and this implementation would fail, but the message was "decimal literal empty" which was confusing)

The change is fairly simple, my main objective is to be able to contribute by starting small.

A few notes:
- To reuse the Span and pattern instances, I match and return the same error values but with a different kind. This way we can keep current and future errors in parse_decimal untouched.
- I also didn't extract a method for the `match` block since the code was small, but if you believe it makes sense, I could also make a `change_error_kind(From, To, self.parse_decimal())` method or pass the desired error kind to `parse_decimal(ErrorKindWhenMissing)`, but I felt the small duplication made it more obvious to read.
- I used `RepetitionQuantifierDecimalMissing` even though it also throws when the decimal is _invalid_, because the current implementation doesn't differentiate missing and invalid decimals, and I didn't want to change the behavior. It is also valid to have a missing decimal in cases like `x{1,}` or `x{,2}`, so I'm sure there's a better name for it.

I also noticed there was no integration tests for error messages. I excluded it from this PR but if you believe it can be valuable (it helped me make sure that my change actually worked) I can include it in this PR as a separate commit (or a different PR if you prefer). The test looks like this:

```rust
fn repetition_quantifier_decimal_error() {
    assert_panic_message(r"\\u{[^}]*}", r#"
regex parse error:
    \\u{[^}]*}
        ^
error: decimal literal empty
"#);
```

I'm ready to do my homework if I missed something, so just let me know!

BurntSushi commented on regex-syntax/src/ast/mod.rs:
```diff
@@ -307,6 +310,9 @@ impl fmt::Display for ErrorKind {
                            the start must be <= the end")
             }
+            RepetitionQuantifierDecimalMissing => {
+                write!(f, "Repetition quantifier expects a valid decimal")
```
Could this match the style of surrounding style of error messages? i.e., Start with a lowercase letter.

BurntSushi commented on regex-syntax/src/ast/parse.rs:
```diff
@@ -1113,7 +1113,15 @@ impl<'s, P: Borrow<Parser>> ParserI<'s, P> {
         }
-        let count_start = self.parse_decimal()?;
+        let count_start = match self.parse_decimal() {
+            Err(ast::Error {kind: ast::ErrorKind::DecimalEmpty, span, pattern}) => return Err(ast::Error {
```
Could you make this code match the style of the surrounding code? i.e., Keep all lines limited to 79 columns inclusive. Please fix this for any other long lines introduced in this PR. (It is hard to see them in Github's PR diff view.)

It might be clearest to "fix" these long lines by extracting this into a separate method, as you suggested. Then you can reuse that below.

BurntSushi commented on regex-syntax/src/ast/mod.rs:
```diff
@@ -153,6 +153,9 @@ pub enum ErrorKind {
     RepetitionCountInvalid,
+    /// An opening `{` was not followed by a valid decimal value.
+    /// For example, `x{}` or `x{]}` would fail.
+    RepetitionQuantifierDecimalMissing,
```
With this change, it looks like `DecimalEmpty` is no longer used. This is a bit weird, but I see why it makes sense to do that. Could you update the docs of `DecimalEmpty` to say something like:

```rust
/// Note that this error variant is no longer used. Namely, a decimal
/// number can only appear as a repetition quantifier. When the number
/// in a repetition quantifier is empty, then it gets its own specialized
/// error, `RepetitionQuantifierDecimalMissing`.

BurntSushi replied:
Also, could you name this `RepetitionCountDecimalEmpty` in order to be consistent with other error variants?

christianrondeau commented:
Changes:
- Fixed upper case letter in error message (good catch)
- Extracted a function for the error specialization. Called it `specialize_err` and made it generic, so it's simple to reuse and read. It's now under 79 columns wide.
- Updated the doc of DecimalEmpty as asked (I think your text was perfect)
- Renamed `RepetitionQuantifierDecimalMissing ` to `RepetitionCountDecimalEmpty`
- Since the code is already a little bit more modified than I initially wanted, I included my integration test, I it wil be easier for you to tell me if you wish to keep it or if I should remove it.

For your first question, the code currently doesn't handle `a{}` and `a{]}` differently. So "empty" is technically correct, though I thought "missing" was more precise since it "expected" one, but didn't found one (i.e. it's not empty, it's just not there). But my thought process is probably not valid for a future reader, so I used your change. Showing a different error message for empty or invalid decimal would require looking further just for the purposes of the error message, so I thought it wasn't worth the complexity.

BurntSushi commented on regex-syntax/src/ast/parse.rs:
```diff
@@ -1129,7 +1133,11 @@ impl<'s, P: Borrow<Parser>> ParserI<'s, P> {
+                    self.parse_decimal(),
+                    ast::ErrorKind::DecimalEmpty,
+                    ast::ErrorKind::RepetitionCountDecimalEmpty,
+                )?;
```
Ah yes, this is much nicer. I like `specialize_err`. :-)

CAD97 commented on regex-syntax/src/ast/parse.rs:
```diff
@@ -2260,6 +2268,29 @@ impl<'p, 's, P: Borrow<Parser>> ast::Visitor for NestLimiter<'p, 's, P> {
+    from: ast::ErrorKind,
+    to: ast::ErrorKind,
+) -> Result<T> {
+    if let Err(e) = result {
```
This could be written with a match guard:

```rust
match result {
    Err(e) if e.kind == from => Err(ast::Error { kind: to, ..e }),
    _ => result,
}
```

CAD97 replied:
Wait nevermind this is moving
